File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/coroutines/parallel/java/DashboardService.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.coroutines.parallel.java;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;

// DashboardService combines data from multiple async sources.
public class DashboardService {

    private final UserApiClient userApiClient;
    private final OrderApiClient orderApiClient;
    private final ExecutorService executorService;

    public DashboardService(UserApiClient userApiClient,
                            OrderApiClient orderApiClient,
                            ExecutorService executorService) {

        this.userApiClient = userApiClient;
        this.orderApiClient = orderApiClient;
        this.executorService = executorService;
    }

    public CompletableFuture<Dashboard> loadDashboard(String userId) {

        // Starts loading the user asynchronously.
        CompletableFuture<User> userFuture =
                CompletableFuture.supplyAsync(
                        () -> userApiClient.fetchUserById(userId),
                        executorService
                );

        // Starts loading orders asynchronously.
        CompletableFuture<List<Order>> ordersFuture =
                CompletableFuture.supplyAsync(
                        () -> orderApiClient.fetchOrdersByUserId(userId),
                        executorService
                );

        // thenCombine waits for both async operations to complete
        // and combines the results into a Dashboard object.
        return userFuture.thenCombine(
                ordersFuture,
                Dashboard::new
        );
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/coroutines/parallel/kotlin/DashboardService.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.coroutines.parallel.kotlin

import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

class DashboardService(
    private val userApiClient: UserApiClient,
    private val orderApiClient: OrderApiClient
) {

    suspend fun loadDashboard(userId: String): Dashboard = coroutineScope {
        // async starts a coroutine for loading the user.
        val userDeferred = async {
            userApiClient.fetchUserById(userId)
        }

        // async starts another coroutine for loading orders.
        val ordersDeferred = async {
            orderApiClient.fetchOrdersByUserId(userId)
        }

        // await waits for the results.
        // If one coroutine fails, coroutineScope cancels the others.
        Dashboard(
            user = userDeferred.await(),
            orders = ordersDeferred.await()
        )
    }
}